JDBI 3 Extensions

qualidafial opened the pull request:
## API changes

* Added `ExtensionFactory`, `ExtensionConfig`, `ExtensionCallback`, `ExtensionConsumer` interfaces
* `DBI` method `open(Class)` replaced with `withExtension(Class, ExtensionCallback)` and `useExtension(Class, ExtensionConsumer)`
* Added `DBI` methods `registerCollectorFactory(CollectorFactory)`,  `registerExtensionFactory(ExtensionFactory)`, and `configureExtension(Class<C>, Consumer<C>)`.
* `DBI.onDemand` restrictions: only public interface types are supported for now--allows us to use proxies without bringing in third party dependencies. The handle from an on-demand extension method call is _always_ closed as the method returns, which means that SQL object methods that return iterables, streams, or other SQL objects (`@CreateSqlObject`) will throw an error when you try to access the returned value, as the handle is closed. (See #271)
* Added method `JdbiPlugin.customizeDbi(DBI)` method. This was necessary so that extensions get installed on the DBI, making those extensions available for use with `DBI.onDemand(Class)`.
* Added `Handle` methods `registerExtension(ExtensionFactory)` and `configureExtension(Class, Consumer)`.
* Interface `CloseMe` removed. Extensions (including SQL objects) are no longer `AutoCloseable`. Users manage the handle resource explicitly and attach extensions to them, or use callback methods like `withExtension` to have the handle managed for them.

(Comprises #295, #271)

To do:

* [x] Finish generic `onDemand` implementation in core
  * [x] Lazy init handle so `onDemand` extensions only create handle when they need to (`equals`, `hashCode`, `toString` being counterexamples).
  * [x] Redirect tests to `DBI.onDemand`
  * [x] Revise tests that assume `onDemand` extensions can provide streams or fluent iterators, to expect the opposite now.
* [x] Replace `HandleDing` with `Supplier<Handle>`, and stop doing handle retain/release in sqlobject.
* [x] Move static configuration in SqlObjectBuilder into config class
* ~~Get rid of CallbackFailedException in favor of exception transparency in callback interfaces, while I'm thinking about it~~ Oh hell no
* [x] Remove method handlers for `AutoCloseable` and friends.
* [x] Remove CloseHandler
* [x] GetHandle's `useHandle` and `withHandle` could be default methods, and could remove cglib handlers `UseHandleHandler` and `WithHandleHandler`

qualidafial commented:
@stevenschlansker @brianm This pull request is ready to look at.

![](https://cdn.meme.am/instances/37052206.jpg)

qualidafial commented:
@stevenschlansker Any chance I could get a code review on this one?

stevenschlansker commented:
Yeah sorry, I've been slammed at work... will try to get to all the PRs this week.

> On Apr 7, 2016, at 9:38 AM, Matthew Hall <notifications@github.com> wrote:
> 
> @stevenschlansker Any chance I could get a code review on this one?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>

stevenschlansker commented on core/src/main/java/org/jdbi/v3/DBI.java:
```diff
@@ -370,41 +382,100 @@ public void useTransaction(final TransactionIsolationLevel isolation, final Tran
+        });
+    }
+
+    private class LazyHandle implements Supplier<Handle>, AutoCloseable {
```
Probably worth making a standalone class?

qualidafial replied:
Wouldn't hurt. Do you have any reuse cases in mind?

stevenschlansker replied:
Nope.  Just struck me as not related enough to the enclosing class to be inner.

stevenschlansker commented on core/src/main/java/org/jdbi/v3/DBI.java:
```diff
@@ -370,41 +382,100 @@ public void useTransaction(final TransactionIsolationLevel isolation, final Tran
+        if (!Modifier.isPublic(extensionType.getModifiers())) {
+            throw new IllegalArgumentException("On-demand extensions types must be public.");
+        }
+        if (!extensionRegistry.hasExtensionFor(extensionType)) {
```
Are you intending to `Optional`-ify this "check then use"?

qualidafial replied:
We do have `ExtensionRegistry.findExtensionFor(Class, Supplier<Handle>)` that returns an `Optional`, but that would require allocating a handle now instead of on-demand.

In this case I'm just verifying that the extension will be available when future calls to the proxy require it.

stevenschlansker commented on core/src/main/java/org/jdbi/v3/ExtensionRegistry.java:
```diff
@@ -0,0 +1,90 @@
+                .filter(entry -> entry.factory.accepts(extensionType))
+                .map(entry -> entry.factory.attach(extensionType, entry.config, handle))
+                .map(extensionType::cast)
+                .findFirst();
```
Should this sanity check that there are no additional elements?  i.e. `findOnly`

qualidafial replied:
This registry follows a last-registered factory wins rule, similar to the other registries. If users need to ensure that a certain factory wins, they must explicitly register the factory (or plugin) last.

stevenschlansker commented on core/src/main/java/org/jdbi/v3/ExtensionRegistry.java:
```diff
@@ -0,0 +1,90 @@
+import org.jdbi.v3.extension.ExtensionFactory;
+import org.jdbi.v3.extension.NoSuchExtensionException;
+
+class ExtensionRegistry {
```
I worry that this class isn't sufficiently threadsafe -- if you publish a config from one thread, there's no guarantees any other thread would observe that config?

stevenschlansker commented on sqlobject/src/main/java/org/jdbi/v3/sqlobject/SqlObject.java:
```diff
@@ -13,236 +13,11 @@
-        else {
-            return q.value();
-        }
+public class SqlObject implements ExtensionConfig<SqlObject> {
```
Should there be a built-in `NoConfig` or the like for extensions which have no configuration?

qualidafial replied:
I thought about that, but decided against it for forward compatibility. Over time a user would expect the config class to stay in the same place, but accumulate more options.

I'm afraid if we provide a `NoConfig` we'd be encouraging extension authors into a situation where they have to break backward compatibility in order to make their extensions configurable.

I'm still debating whether having the configuration class separate from the extension factory is really necessary.

stevenschlansker replied:
Excellent point.

stevenschlansker commented on sqlobject/src/main/java/org/jdbi/v3/sqlobject/SqlObjectFactory.java:
```diff
@@ -0,0 +1,174 @@
+        ResolvedTypeWithMembers d = mr.resolve(extension_type, null, null);
+        for (ResolvedMethod method : d.getMemberMethods()) {
+            Method rawMethod = method.getRawMember();
+            if (rawMethod.isAnnotationPresent(SqlQuery.class)
```
This hardcoded list sucks.  Maybe a meta-annotation?
```java
@SqlObjectAnnotation
public @interface SqlQuery { ... }
```

qualidafial replied:
I could get behind that.

qualidafial replied:
We could go a step further and make `@SqlObjectAnnotation` take a factory class as a parameter, similar to `@SqlStatementCustomizingAnnotation`. Then users are free to introduce their own method handlers.

qualidafial replied:
Hmm, if we give the meta-annotation a factory class, then we expose the Handler interface and by extension cglib's MethodProxy type into the API. The MethodProxy argument is still used in `PassThroughHandler`, so I can't remove it from the interface method just yet.

I'm going to call it `@SqlMethodAnnotation` and just annotate the other annotation classes for now. I'll open an issue to add the rest later.

qualidafial replied:
Another idea to explore: modify the mixins to use meta-annotations

    public interface GetHandle
    {
      @SqlMethod(GetHandleHandler.class)
      Handle getHandle();
    }

Then the mixins go through the same annotation lookup procedure as e.g. `@SqlQuery`, `@SqlBatch`, etc.

stevenschlansker commented on sqlobject/src/main/java/org/jdbi/v3/sqlobject/SqlObjectFactory.java:
```diff
@@ -0,0 +1,174 @@
+            for (final ResolvedMethod method : d.getMemberMethods()) {
+                final Method raw_method = method.getRawMember();
+
+                if (raw_method.isAnnotationPresent(SqlQuery.class)) {
```
The meta-annotation could drive this dispatch as well

stevenschlansker replied:
```java
@SqlObjectAnnotation(QueryHandler.class)
public @interface SqlQuery { ... }
```

stevenschlansker commented on sqlobject/src/main/java/org/jdbi/v3/sqlobject/mixins/GetHandle.java:
```diff
@@ -41,7 +41,14 @@
+            return callback.withHandle(getHandle());
+        }
+        catch (Exception e) {
+            throw new CallbackFailedException(e);
```
This should be exception-transparent like the other PR, yeah?

qualidafial replied:
This is already such a big pull request, and the cognitive load was getting too high already. That's why I opened the separate PR for the exception transparency stuff.

This change was just about migrating from `GetHandleHandler` to a default method.

stevenschlansker commented:
It looks like the Extension interface essentially requires you to use cglib.
I don't suppose there would be any way to tease these apart, so we handle cglib under the hood, and your Extension doesn't need to know / care that we use it?  I only ask because it ties us to cglib, and we couldn't switch to e.g. a MethodHandle or Proxy based approach without breaking API compatibility...

stevenschlansker commented:
Dude this PR rocks!  I'm excited.  SqlObject being tied to core so intimately was one of my pet peeves, with `HandleDing` being second on the list.  Sending buckets of internet karma your way :smile: 

Will let you do one more pass, poke me when you want a re-review or merge.

qualidafial commented:
> It looks like the Extension interface essentially requires you to use cglib.

There's no cglib in core, just sqlobject. Am I missing something?

stevenschlansker commented:
> On Apr 8, 2016, at 4:13 PM, Matthew Hall <notifications@github.com> wrote:
> 
> It looks like the Extension interface essentially requires you to use cglib.
> 
> There's no cglib in core, just sqlobject. Am I missing something?


Sorry, it's late on Friday and I can't read!

qualidafial commented:
I think I've addressed most of your concerns. I'd like to hold off on adding factories to the meta-annotation for now--I feel like that's somewhat orthogonal, and should live in a separate issue.

qualidafial commented:
Ready for another review

qualidafial commented:
@stevenschlansker ping

qualidafial commented:
Actually, now I'm working on making the annotations specify a `HandlerFactory` class right now. Should be done in ~30min

qualidafial commented:
@stevenschlansker Ready for review

stevenschlansker commented:
What a mouthful!
